pkg/packagekit/kpatch.jsx

```diff
@@ -0,0 +1,319 @@
+
+        this.checkSetup = this.checkSetup.bind(this);
+        this.handleChange = this.handleChange.bind(this);
+        this.handleService = this.handleService.bind(this);
+        this.handleInstall = this.handleInstall.bind(this);
+        this.installPatches = this.installPatches.bind(this);
+    }
+
+    componentDidMount() {
+        check_missing_packages(["kpatch", "kpatch-dnf"])
+                .then(d => {
+                    this.setState({
+                        loaded: true,
+                        unavailable: d.unavailable_names || [],
+                        missing: d.missing_names || [],
+                    });
+
+                    if ((d.unavailable_names || []).length === 0 && (d.missing_names || []).length === 0)
+                        this.checkSetup();
+                })
+                .catch(e => {
+                    console.log("Failed to query info about availablity of kpatch");
+                });
+
+        this.kpatchService.addEventListener('changed', () => {
+            this.setState({ enabled: this.kpatchService.enabled, futureEnabled: this.kpatchService.enabled });
+        });
+    }
+
+    handleService(enable) {
+        if (enable)
+            return this.kpatchService.enable();
+        else
+            return this.kpatchService.disable();
+    }
+
+    installPatches(auto) {
+        return cockpit.spawn(["dnf", "-y", "kpatch", auto ? "auto" : "manual"], { superuser: "require", err: "message" });
+    }
+
+    checkSetup() {
+        return new Promise((resolve, reject) => {
+            cockpit.file("/etc/dnf/plugins/kpatch.conf").read()
+                    .then(data => {
+                        if (!data) {
+                            resolve();
+                        } else {
+                            const autoupdate = data.split("\n").find(l => l.indexOf("autoupdate") >= 0);
+                            const auto = autoupdate.endsWith("True");
+                            this.setState({ auto: auto, futureAuto: auto }, resolve);
+                        }
+                    })
+                    .catch(reject);
+        });
+    }
+
+    handleInstall() {
+        this.setState({ updating: true });
+        install_dialog(this.state.missing)
+                .then(() => {
+                    this.setState({ missing: [], updating: false });
+                })
+                .catch(() => this.setState({ updating: false }));
+    }
+
+    handleChange() {
+        this.setState({ updating: true });
+
+        const promises = [];
+
+        if (this.state.auto !== this.state.futureAuto)
+            promises.push(this.installPatches(this.state.futureAuto));
+
+        if (this.state.enabled !== this.state.futureEnabled)
+            promises.push(this.handleService(this.state.futureEnabled));
+
+        Promise.all(promises)
+                .then(() => this.setState({ showModal: false }))
+                .catch(console.log)
```

Comment:
Let's JSON.stringify the exception at least? Otherwise we don't see it in the test logs.